cobalt/build/build_info.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
+# Copyright 2023 The Cobalt Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Generates a Cobalt Build Info json."""
+
+import datetime
+import json
+import os
+import re
+import subprocess
+import sys
+
+_FILE_DIR = os.path.dirname(__file__)
+_COMMIT_COUNT_BUILD_ID_OFFSET = 1000000
+
+# git log --grep can't handle complex patterns.
+_BUILD_ID = r'[1-9]' + r'[0-9]' * 6 + r'[0-9]*'
+_GIT_REV = r'[0-9a-f]' * 40
+
+# git log --grep can't handle capture groups.
+_BUILD_ID_WITH_CAPTURE = f'({_BUILD_ID})'
+_GIT_REV_WITH_CAPTURE = f'({_GIT_REV})'
+
+_BUILD_ID_PATTERN = r'^BUILD_NUMBER={}$'
+_GIT_REV_PATTERN = r'^GitOrigin-RevId: {}$'
+
+
+def get_build_id_and_git_rev_from_commits(cwd):
+  # Build id and git rev must come from the same commit.
+  search_pattern = _BUILD_ID_PATTERN.format(_BUILD_ID)
+  output = subprocess.check_output(
+      ['git', 'log', '--grep', search_pattern, '-1', '--pretty=%b'],
+      cwd=cwd).decode()
+
+  # Gets build id.
+  compiled_build_id_pattern = re.compile(
+      _BUILD_ID_PATTERN.format(_BUILD_ID_WITH_CAPTURE), flags=re.MULTILINE)
+  match_build_id = compiled_build_id_pattern.search(output)
+
+  if match_build_id is None:
+    return None
```

Comment:
This should probably return `(None, None)` or `('', '')` instead. That makes the unpacking easier below.